packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,573 @@
+          continue;
+        }
+        if (!isPersistentKey(stateOrPersistentKey)) {
+          if (firstState === null) {
+            firstState = stateOrPersistentKey;
+          }
+
+          // amongst states of the same epoch, choose the one with the same view of reloadedCp
+          if (
+            reloadedCpSlot < stateOrPersistentKey.slot &&
+            toHexString(getBlockRootAtSlot(stateOrPersistentKey, reloadedCpSlot)) === reloadedCp.rootHex
+          ) {
+            return stateOrPersistentKey;
+          }
+        }
+      }
+    }
+
+    return firstState;
+  }
+
+  clear(): void {
+    this.cache.clear();
+    this.epochIndex.clear();
+  }
+
+  /** ONLY FOR DEBUGGING PURPOSES. For lodestar debug API */
+  dumpSummary(): routes.lodestar.StateCacheItem[] {
+    return Array.from(this.cache.keys()).map((key) => {
+      const cp = fromCheckpointKey(key);
+      const stateOrPersistentKey = this.cache.get(key);
+      return {
+        slot: computeStartSlotAtEpoch(cp.epoch),
+        root: cp.rootHex,
+        reads: this.cache.readCount.get(key) ?? 0,
+        lastRead: this.cache.lastRead.get(key) ?? 0,
+        checkpointState: true,
+        persistentKey:
+          stateOrPersistentKey !== undefined && isPersistentKey(stateOrPersistentKey)
+            ? stateOrPersistentKey
+            : undefined,
+      };
+    });
+  }
+
+  /** ONLY FOR DEBUGGING PURPOSES. For spec tests on error */
+  dumpCheckpointKeys(): string[] {
+    return Array.from(this.cache.keys());
+  }
+
+  /**
+   * Delete all items of an epoch from disk and memory
+   */
+  private async deleteAllEpochItems(epoch: Epoch): Promise<void> {
+    for (const rootHex of this.epochIndex.get(epoch) || []) {
+      const key = toCheckpointKey({rootHex, epoch});
+      const stateOrPersistentKey = this.cache.get(key);
+      if (stateOrPersistentKey !== undefined && isPersistentKey(stateOrPersistentKey)) {
+        await this.persistentApis.remove(stateOrPersistentKey);
+        this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.pruneFinalized});
+      }
+      this.cache.delete(key);
+    }
+    this.epochIndex.delete(epoch);
+  }
+}
+
+export function toCheckpointHex(checkpoint: phase0.Checkpoint): CheckpointHex {
+  return {
+    epoch: checkpoint.epoch,
+    rootHex: toHexString(checkpoint.root),
+  };
+}
+
+export function toCheckpointKey(cp: CheckpointHex | phase0.Checkpoint): CheckpointKey {
+  if (isCheckpointHex(cp)) {
+    return `${cp.rootHex}_${cp.epoch}`;
+  }
+  return `${toHexString(cp.root)}_${cp.epoch}`;
+}
+
+export function fromCheckpointKey(key: CheckpointKey): CheckpointHex {
+  const [rootHex, epoch] = key.split("_");
+  return {
+    rootHex,
+    epoch: Number(epoch),
+  };
+}
+
+function isPersistentKey(
+  stateOrPersistentKey: CachedBeaconStateAllForks | PersistentKey
+): stateOrPersistentKey is PersistentKey {
+  return (stateOrPersistentKey as CachedBeaconStateAllForks).epochCtx === undefined;
```

Comment:
Consider wrapping the items in the cache in an enum style to have a cleaner separation of types:

```ts
type CacheItem = {type: "persistent", value: X} | {type: "in-memory", value: Y}
```